### PR TITLE
feat(clickhouse): add support for decimal(x,y), decimal32(x) and match

### PR DESCRIPTION
**Pull request by ogirardot**
### Brief summary of the change made
Add support for clickhouse special decimals and match(regexp) functions, it includes : 
```
Numeric(x, y)
Decimal(x, y)
Decimal32(x)
Decimal64(x)
Decimal128(x)
Decimal256(x)
match(field, 'regexp')
```

### Are there any other side effects of this change that we should be aware of?
Nope

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 1]
```diff
@@ -350,6 +350,32 @@ class DatatypeSegment(BaseSegment):
...
+                    StringParser("DECIMAL32", CodeSegment, type="data_type_identifier"),
+                    StringParser("DECIMAL64", CodeSegment, type="data_type_identifier"),
+                    StringParser(
+                        "DECIMAL128", CodeSegment, type="data_type_identifier"
+                    ),
+                    StringParser(
+                        "DECIMAL256", CodeSegment, type="data_type_identifier"
+                    ),
+                ),
+                OptionallyBracketed(Ref("NumericLiteralSegment")),  # precision
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:375]: Should this be `Optional(Bracketed(...))` rather than `OptionallyBracketed`?

Comment (ogirardot) [reply src/sqlfluff/dialects/dialect_clickhouse.py:375]: I'm not sure about the difference, there are already quite a few OptionallyBracketed in the file, should it be refactored ?

Comment (alanmcruickshank) [reply src/sqlfluff/dialects/dialect_clickhouse.py:375]: `OptionallyBracketed` implies that something can be bracketed or not bracketed - but one of them must be there. e.g. `OptionallyBracketed("FOO")` would match `FOO` and `(FOO)`.

`Bracketed("FOO", optional=True)` requires brackets around `FOO`, but also allows nothing to be present. i.e. matches `(FOO)` and ``, but _wouldn't_ match an unbracketed `FOO`.

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 2]
```diff
@@ -350,6 +350,32 @@ class DatatypeSegment(BaseSegment):
...
+                optional=True,
+            ),
+        ),
```
Comment (alanmcruickshank) [inline src/sqlfluff/dialects/dialect_clickhouse.py:378]:
```suggestion
        # double args
        Sequence(
            OneOf(
                StringParser("DECIMAL", CodeSegment, type="data_type_identifier"),
                StringParser("NUMERIC", CodeSegment, type="data_type_identifier"),
            ),
            Ref("BracketedArguments", optional=True),
        ),
        # single args
        Sequence(
            OneOf(
                StringParser("DECIMAL32", CodeSegment, type="data_type_identifier"),
                StringParser("DECIMAL64", CodeSegment, type="data_type_identifier"),
                StringParser(
                    "DECIMAL128", CodeSegment, type="data_type_identifier"
                ),
                StringParser(
                    "DECIMAL256", CodeSegment, type="data_type_identifier"
                ),
            ),
            Bracketed(Ref("NumericLiteralSegment"), optional=True),  # precision
        ),
```

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 1 above)
Comment (ogirardot) [reply src/sqlfluff/dialects/dialect_clickhouse.py:375]: OK then yes, and taking a second look at this it's not even right : 
```
>> SELECT 1::Decimal32

SELECT CAST('1', 'Decimal32')

Query id: 72b396fa-10d6-44fd-9771-229fb886c7a3


Elapsed: 0.000 sec. 

Received exception:
Code: 42. DB::Exception: Decimal32 | Decimal64 | Decimal128 | Decimal256 data type family must have exactly one arguments: scale: In scope SELECT CAST('1', 'Decimal32'). (NUMBER_OF_ARGUMENTS_DOESNT_MATCH)
```

So Decimal32() or Decimal32 is incorrect so the optional part is not even true.
Let me fix this

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 2 above)
Comment (ogirardot) [reply src/sqlfluff/dialects/dialect_clickhouse.py:378]: Thanks, I'll restart from this because taking a second look Decimal32 is not even allowed

**Pull request merged**